test/unit/puppet_fact_parser_test.rb

```diff
@@ -438,8 +438,362 @@ def setup
+
+          test "#dmi_board_product" do
+            if facterversion.to_f == '3.0'
+              assert_kind_of String, subject.send(:dmi_board_product)
+            else
+              assert_nil subject.send(:dmi_board_product)
+            end
+          end
+
+          test "#architecture_fact" do
+            assert_includes ['sun4v', 'i86pc'], subject.send(:architecture_fact)
+          end
+
+          test "#hardware_isa" do
+            assert_includes ['sparc', 'i386'], subject.send(:hardware_isa)
+          end
+        end
+      end
+    end
+
+    describe 'Windows 2012' do
+      let(:os_name) { 'windows' }
+      let(:os_major) { '2012' }
+
+      ['2.2', '3.0', '3.14'].each do |facterversion|
+        describe "Facter #{facterversion}" do
+          let(:facterversion) { facterversion }
+
+          test "#os_name" do
+            assert_equal 'windows', subject.send(:os_name)
+          end
+
+          test "#os_release" do
+            assert_match(/^6\.2\.\d+$/, subject.send(:os_release))
+          end
+
+          test "#architecture" do
+            refute_nil subject.architecture
+            assert_not_equal 'amd64', subject.architecture.name
+            assert_includes ['i386', 'x64'], subject.architecture.name
+          end
+
+          test "#distro_id" do
+            # Based on LSB (Linux Standard Base) but Windows isn't Linux
+            assert_nil subject.send(:distro_id)
+          end
+
+          test "#distro_codename" do
+            # Based on LSB (Linux Standard Base) but Windows isn't Linux
+            assert_nil subject.send(:distro_codename)
+          end
+
+          test "#distro_description" do
+            # Based on LSB (Linux Standard Base) but Windows isn't Linux
+            assert_nil subject.send(:distro_description)
+          end
+
+          test "#dmi_product_name" do
+            if facterversion.to_f >= 2.2
+              assert_kind_of String, subject.send(:dmi_product_name)
+            else
+              assert_nil subject.send(:dmi_product_name)
+            end
+          end
+
+          test "#dmi_board_product" do
+            # No such thing on Windows
+            assert_nil subject.send(:dmi_board_product)
+          end
+
+          test "#architecture_fact" do
+            assert_equal 'x64', subject.send(:architecture_fact)
+          end
+
+          test "#hardware_isa" do
+            assert_equal 'x64', subject.send(:architecture_fact)
+          end
+        end
+      end
+    end
+  end
+
   private
 
+  def get_facterdb_facts(facterversion, os_name, os_major)
+    require 'facterdb'
+    # This uses the legacy facts since it's always present
+    filter = "facterversion=/^#{Regexp.escape(facterversion)}\./ and operatingsystem=#{os_name} and operatingsystemmajrelease=#{os_major}"
+    result = FacterDB.get_facts(filter)
+    raise "No facts found for #{os_name} #{os_major} on Facter #{facterversion}" if result.empty?
+    result.first.dup
+  end
```

Comment:
This method needs to have a LRU cache - when its called with the very same arguments, it should promptly return the result. Now, ideally it should be stored in the test instance as a member attribute, but I do not know if spec/minitest does not create new class for each individual test. Then in that case use a global variable or a class attribute. Please do not cache all possible combinations, that will take too much memory, simply "memorize" the last input and output.

My second concern is FacterDB's `@database` field. It concatenates huge amount of JSON files into a string in-memory. If I was designing this, I would probably move this to the build phase creating one huge JSON file, so it can only be loaded during runtime, preferably memory-mapped. Anyway, back to the problem, it claims 15MB of memory and it never returns it back. This means after this test suite is done, 15MB of memory is lost until the test process dies and OS reclaims it. Please implement a `after(:suite)` block that will set the class attribute to nil so Ruby VM can use that memory again.

You know what - I have an idea: https://github.com/voxpupuli/facterdb/pull/171

Reply:
I agree and implemented a similar cache in https://github.com/voxpupuli/rspec-puppet-facts/commit/43bb17e9f894421daef73e833ca3ea0a506f8d9c but we can't really use that here.